### PR TITLE
Fix DynRankView performance by re-introducing shortcut operators

### PR DESCRIPTION
**Pull request by crtrott**
This addresses perf issues documented in #7604

### CODE REVIEW

**Comment by crtrott** [comment]
I suspect that some compilers may want the unreachable thing again since all the returns happen inside of is constexpr expressions. But let's see what happens.

**Review by lucbv — APPROVED** [review]
Looks reasonable to me

File: containers/src/Kokkos_DynRankView.hpp [hunk 1]
```diff
@@ -574,10 +577,97 @@ class DynRankView : private View<DataType*******, Properties...> {
...
+      if constexpr (is_layout_left) {
+        return view_type::data()[i0 + view_type::stride(1) *
+                                          (i1 + i2 * view_type::extent(1))];
+      } else if constexpr (is_layout_right) {
+        return view_type::data()[(i0 * view_type::extent(1) + i1) *
+                                     view_type::extent(2) +
+                                 i2];
+      } else {
+        return view_type::data()[i0 * view_type::stride(0) +
+                                 i1 * view_type::stride(1)];
```
Comment (masterleinad) [inline containers/src/Kokkos_DynRankView.hpp:659]:
```suggestion
        return view_type::data()[i0 * view_type::stride(0) +
                                 i1 * view_type::stride(1) +  i2 * view_type::stride(2)];
```

**Comment by masterleinad** [comment]
> I suspect that some compilers may want the unreachable thing again since all the returns happen inside of is constexpr expressions. But let's see what happens.

We should probably have @ndellingwood check since we don't test with the `Intel` classic compiler otherwise.

File: containers/src/Kokkos_DynRankView.hpp (hunk 1 above)
Comment (lucbv) [reply containers/src/Kokkos_DynRankView.hpp:659]: The fact that this one did not make the SNL build fail probably hints at a lack of coverage? Are there layout stride tests for this?

Comment (crtrott) [reply containers/src/Kokkos_DynRankView.hpp]: pushed the fix

**Review by masterleinad — APPROVED** [review]

File: containers/src/Kokkos_DynRankView.hpp [hunk 2]
```diff
@@ -574,10 +577,98 @@ class DynRankView : private View<DataType*******, Properties...> {
...
 
-  KOKKOS_FUNCTION reference_type
-  operator()(index_type i0 = 0, index_type i1 = 0, index_type i2 = 0,
-             index_type i3 = 0, index_type i4 = 0, index_type i5 = 0,
-             index_type i6 = 0) const {
+  // Adding shortcut operators for rank-0 to rank-3 for default layouts
+  // and access modalities.
+  // This removes performance overhead for always using rank-7 mapping.
+  // See https://github.com/kokkos/kokkos/issues/7604
+  // When boundschecking is enabled we still go through the underlying
```
Comment (dalg24) [inline containers/src/Kokkos_DynRankView.hpp:584]:
```suggestion
  // When bounds checking is enabled, we still go through the underlying
```

File: containers/src/Kokkos_DynRankView.hpp [hunk 3]
```diff
@@ -574,10 +577,98 @@ class DynRankView : private View<DataType*******, Properties...> {
...
+  // rank-7 View to leverage the error checks there.
+
+  KOKKOS_FUNCTION reference_type operator()() const {
+#ifdef KOKKOS_ENABLE_DEBUG
+    if (rank() != 0u)
+      Kokkos::abort(
+          "DynRankView operator() called with invalid number of arguments.");
```
Comment (dalg24) [inline containers/src/Kokkos_DynRankView.hpp:591]: You could also do
```C++
KOKKOS_EXPECTS(rank() != 0u && "DynRankView operator() called with invalid number of arguments");

**Comment by ndellingwood** [comment]
I tested this with intel/19 with Wall and Werror, no issues showed up

```
cmake -DCMAKE_CXX_COMPILER=icpc -DCMAKE_CXX_FLAGS="-Wall -Werror -diag-disable=1011 -diag-disable=869" -DKokkos_ENABLE_SERIAL=ON -DKokkos_ARCH_BDW=ON -DKokkos_ENABLE_TESTS=ON -DKokkos_ENABLE_EXAMPLES=OFF -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_CXX_EXTENSIONS=OFF -DCMAKE_BUILD_TYPE=RELEASE -DKokkos_ENABLE_DEPRECATION_WARNINGS=OFF -DKokkos_ENABLE_DEPRECATED_CODE_4=OFF $KOKKOS_PATH
```

File: containers/src/Kokkos_DynRankView.hpp (hunk 3 above)
Comment (crtrott) [reply containers/src/Kokkos_DynRankView.hpp:591]: was copied from the `operator[]` do you want me to change it everywhere?

Comment (crtrott) [reply containers/src/Kokkos_DynRankView.hpp:591]: Talked with Damien, will change in a follow up everywhere.

**Pull request merged**